{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a721dcd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "import pandas as pd\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8acfdb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FILENAME = \"/Users/anaya/Downloads/full database.xml\"\n",
    "FILENAME = \"/home/anaya/Projects/deep-learning-26wi/datasets/drugbank.xml\"\n",
    "OUT_CSV_NAME = \"/home/anaya/Projects/deep-learning-26wi/datasets/interactions.csv\"\n",
    "OUT_TSV_NAME = \"/home/anaya/Projects/deep-learning-26wi/datasets/interactions.tsv\"\n",
    "NAMESPACE = {\"db\": \"http://www.drugbank.ca\"}\n",
    "tree = ET.parse(FILENAME)\n",
    "root = tree.getroot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1fade11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0611a4352b9140f2ba05ebdfc01a68c9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/19830 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with open(OUT_TSV_NAME, \"w\") as f:\n",
    "    # f.write(f\"drug_primary_id,drug_name,other_drug_id,other_drug_name,interaction_description\\n\")\n",
    "    f.write(f\"drug_primary_id\\tdrug_name\\tother_drug_id\\tother_drug_name\\tinteraction_description\\n\")\n",
    "    for drug in tqdm(root.findall(\"db:drug\", NAMESPACE)):\n",
    "        drug_name = drug.find(\"db:name\", NAMESPACE).text\n",
    "        drug_id = drug.find(\"db:drugbank-id[@primary='true']\", NAMESPACE).text\n",
    "        drug_interactions = drug.find(\"db:drug-interactions\", NAMESPACE)\n",
    "        for interaction in drug_interactions.findall(\"db:drug-interaction\", NAMESPACE):\n",
    "            other_drug_name = interaction.find(\"db:name\", NAMESPACE).text\n",
    "            other_drug_id = interaction.find(\"db:drugbank-id\", NAMESPACE).text\n",
    "            interaction_description = interaction.find(\"db:description\", NAMESPACE).text\n",
    "            # f.write(f\"{drug_id},{drug_name},{other_drug_id},{other_drug_name},{interaction_description}\\n\")\n",
    "            f.write(f\"{drug_id}\\t{drug_name}\\t{other_drug_id}\\t{other_drug_name}\\t{interaction_description}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2c1342e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "for inter in root.iter(\"drug-interaction\"):\n",
    "    print(\"hi\")\n",
    "    print(inter.attrib)\n",
    "    i += 1\n",
    "    if i > 10:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8ae20fbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "def get_all_xml_tags(filename):\n",
    "    \"\"\"\n",
    "    Reads an XML file and returns a unique list of all element tags.\n",
    "\n",
    "    Args:\n",
    "        filename (str): The path to the XML file.\n",
    "\n",
    "    Returns:\n",
    "        list: A sorted list of unique tag names.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Parse the XML file\n",
    "        tree = ET.parse(filename)\n",
    "        root = tree.getroot()\n",
    "\n",
    "        # Use a set comprehension with iter() to get all unique tags\n",
    "        tags = {elem.tag for elem in root.iter()}\n",
    "\n",
    "        # Convert to a list and sort for a consistent output\n",
    "        return sorted(list(tags))\n",
    "    except ET.ParseError as e:\n",
    "        print(f\"Error parsing XML file: {e}\")\n",
    "        return []\n",
    "    except FileNotFoundError:\n",
    "        print(f\"Error: The file '{filename}' was not found.\")\n",
    "        return []\n",
    "\n",
    "# Example Usage:\n",
    "# Assuming you have an 'example.xml' file\n",
    "# tags_list = get_all_xml_tags('example.xml')\n",
    "# print(tags_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "05d93ce1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['absorption',\n",
       " 'action',\n",
       " 'actions',\n",
       " 'adverse-reaction',\n",
       " 'affected-organism',\n",
       " 'affected-organisms',\n",
       " 'ahfs-codes',\n",
       " 'allele',\n",
       " 'alternative-parent',\n",
       " 'amino-acid-sequence',\n",
       " 'approved',\n",
       " 'article',\n",
       " 'articles',\n",
       " 'atc-code',\n",
       " 'atc-codes',\n",
       " 'attachment',\n",
       " 'attachments',\n",
       " 'average-mass',\n",
       " 'calculated-properties',\n",
       " 'carrier',\n",
       " 'carriers',\n",
       " 'cas-number',\n",
       " 'categories',\n",
       " 'category',\n",
       " 'cellular-location',\n",
       " 'chromosome-location',\n",
       " 'citation',\n",
       " 'class',\n",
       " 'classification',\n",
       " 'clearance',\n",
       " 'company',\n",
       " 'cost',\n",
       " 'country',\n",
       " 'defining-change',\n",
       " 'description',\n",
       " 'direct-parent',\n",
       " 'dosage',\n",
       " 'dosage-form',\n",
       " 'dosages',\n",
       " 'dpd-id',\n",
       " 'drug',\n",
       " 'drug-interaction',\n",
       " 'drug-interactions',\n",
       " 'drugbank',\n",
       " 'drugbank-id',\n",
       " 'drugs',\n",
       " 'effect',\n",
       " 'ema-ma-number',\n",
       " 'ema-product-code',\n",
       " 'ended-marketing-on',\n",
       " 'enzyme',\n",
       " 'enzymes',\n",
       " 'experimental-properties',\n",
       " 'expires',\n",
       " 'external-identifier',\n",
       " 'external-identifiers',\n",
       " 'external-link',\n",
       " 'external-links',\n",
       " 'fda-application-number',\n",
       " 'fda-label',\n",
       " 'food-interaction',\n",
       " 'food-interactions',\n",
       " 'form',\n",
       " 'gene-name',\n",
       " 'gene-sequence',\n",
       " 'gene-symbol',\n",
       " 'general-function',\n",
       " 'general-references',\n",
       " 'generic',\n",
       " 'go-classifier',\n",
       " 'go-classifiers',\n",
       " 'group',\n",
       " 'groups',\n",
       " 'half-life',\n",
       " 'id',\n",
       " 'identifier',\n",
       " 'inchikey',\n",
       " 'indication',\n",
       " 'induction-strength',\n",
       " 'ingredients',\n",
       " 'inhibition-strength',\n",
       " 'international-brand',\n",
       " 'international-brands',\n",
       " 'isbn',\n",
       " 'kind',\n",
       " 'kingdom',\n",
       " 'known-action',\n",
       " 'labeller',\n",
       " 'left-element',\n",
       " 'level',\n",
       " 'link',\n",
       " 'links',\n",
       " 'locus',\n",
       " 'manufacturer',\n",
       " 'manufacturers',\n",
       " 'mechanism-of-action',\n",
       " 'mesh-id',\n",
       " 'metabolism',\n",
       " 'mixture',\n",
       " 'mixtures',\n",
       " 'molecular-weight',\n",
       " 'monoisotopic-mass',\n",
       " 'msds',\n",
       " 'name',\n",
       " 'ndc-id',\n",
       " 'ndc-product-code',\n",
       " 'number',\n",
       " 'organism',\n",
       " 'over-the-counter',\n",
       " 'packager',\n",
       " 'packagers',\n",
       " 'patent',\n",
       " 'patents',\n",
       " 'pathway',\n",
       " 'pathways',\n",
       " 'pdb-entries',\n",
       " 'pdb-entry',\n",
       " 'pediatric-extension',\n",
       " 'pfam',\n",
       " 'pfams',\n",
       " 'pharmacodynamics',\n",
       " 'polypeptide',\n",
       " 'price',\n",
       " 'prices',\n",
       " 'product',\n",
       " 'products',\n",
       " 'property',\n",
       " 'protein-binding',\n",
       " 'protein-name',\n",
       " 'pubmed-id',\n",
       " 'reaction',\n",
       " 'reactions',\n",
       " 'ref-id',\n",
       " 'references',\n",
       " 'resource',\n",
       " 'right-element',\n",
       " 'route',\n",
       " 'route-of-elimination',\n",
       " 'rs-id',\n",
       " 'salt',\n",
       " 'salts',\n",
       " 'sequence',\n",
       " 'sequences',\n",
       " 'signal-regions',\n",
       " 'smpdb-id',\n",
       " 'snp-adverse-drug-reactions',\n",
       " 'snp-effects',\n",
       " 'source',\n",
       " 'specific-function',\n",
       " 'started-marketing-on',\n",
       " 'state',\n",
       " 'strength',\n",
       " 'subclass',\n",
       " 'substituent',\n",
       " 'superclass',\n",
       " 'synonym',\n",
       " 'synonyms',\n",
       " 'synthesis-reference',\n",
       " 'target',\n",
       " 'targets',\n",
       " 'textbook',\n",
       " 'textbooks',\n",
       " 'theoretical-pi',\n",
       " 'title',\n",
       " 'toxicity',\n",
       " 'transmembrane-regions',\n",
       " 'transporter',\n",
       " 'transporters',\n",
       " 'unii',\n",
       " 'uniprot-id',\n",
       " 'unit',\n",
       " 'url',\n",
       " 'value',\n",
       " 'volume-of-distribution']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = get_all_xml_tags(FILENAME)\n",
    "[x[len(\"{http://www.drugbank.ca}\"):] for x in l]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64758822",
   "metadata": {},
   "outputs": [],
   "source": [
    "n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "834fe998",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source - https://stackoverflow.com/a/10077069\n",
    "# Posted by K3---rnc, modified by community. See post 'Timeline' for change history\n",
    "# Retrieved 2026-02-21, License - CC BY-SA 3.0\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "def etree_to_dict(t):\n",
    "    d = {t.tag: {} if t.attrib else None}\n",
    "    children = list(t)\n",
    "    if children:\n",
    "        dd = defaultdict(list)\n",
    "        for dc in map(etree_to_dict, children):\n",
    "            for k, v in dc.items():\n",
    "                dd[k].append(v)\n",
    "        d = {t.tag: {k:v[0] if len(v) == 1 else v for k, v in dd.items()}}\n",
    "    if t.attrib:\n",
    "        d[t.tag].update(('@' + k, v) for k, v in t.attrib.items())\n",
    "    if t.text:\n",
    "        text = t.text.strip()\n",
    "        if children or t.attrib:\n",
    "            if text:\n",
    "              d[t.tag]['#text'] = text\n",
    "        else:\n",
    "            d[t.tag] = text\n",
    "    return d\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6693f9f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = ET.parse(FILENAME)\n",
    "as_dict = etree_to_dict(tree.getroot())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "83ba266f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eae87926",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"out.json\", \"w\") as f:\n",
    "    json.dump(as_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a50c63d7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
